cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -91,6 +103,126 @@ func stampedAction(stamp string, action func() error) error {
+// systemd-mount(1)
+type SystemdMountOptions struct {
+	// is tmpfs indicates that "what" should be ignored and a new tmpfs should
+	// be mounted at the location
+	IsTmpfs bool
+	// SurvivesPivotRoot indicates that the mount should persist from the
+	// initramfs to after the pivot_root to normal userspace
+	// this is done by creating systemd unit .wants symlinks on
+	// initrd-device.target in /run
+	// TODO: all current instances of doSystemdMount set this to true, should we
+	// invert or just drop the option and make it the default behavior ?
+	SurvivesPivotRoot bool
+	// NeedsFsck indicates that before returning to the caller, an fsck check
+	// should be performed on the thing being mounted
+	NeedsFsck bool
+	// NoWait will not wait until the systemd unit is active and running, which
+	// is the default behavior
+	NoWait bool
+}
+
+var (
+	doSystemdMount = doSystemdMountImpl
+	counter        = 1
+)
+
+// doSystemdMount will mount "what" at "where" using systemd-mount(1) with
+// various options
+func doSystemdMountImpl(what, where string, opts *SystemdMountOptions) error {
+	if opts == nil {
+		opts = &SystemdMountOptions{}
+	}
+
+	// doesn't make sense to fsck a tmpfs
+	if opts.NeedsFsck && opts.IsTmpfs {
+		return fmt.Errorf("cannot mount %q at %q: impossible to fsck a tmpfs", what, where)
+	}
+
+	whereEscaped := systemd.EscapeUnitNamePath(where)
+	unitName := whereEscaped + ".mount"
+
+	args := []string{"--no-pager", "--no-ask-password", what, where}
+	if opts.IsTmpfs {
+		args = append(args, "--type=tmpfs")
+	}
+
+	if opts.NeedsFsck {
+		args = append(args, "--fsck=yes")
+	}
+
+	// TODO: do we really need this? unclear why we would need to ever not block
+	if opts.NoWait {
+		args = append(args, "--no-block")
+	}
+
+	// note that we do not currently parse any output from systemd-mount, but if
+	// we ever do, take special care surrounding the debug output that systemd
+	// outputs with the "debug" kernel command line present (or equivalently the
+	// SYSTEMD_LOG_LEVEL=debug env var) which will add lots of additional output
+	// to stderr from systemd commands
+	out, err := exec.Command("systemd-mount", args...).CombinedOutput()
+	if err != nil {
+		return osutil.OutputErr(out, err)
+	}
+
+	// finally if it should survive pivot_root() then we need to add symlinks
+	// to /run/systemd
+	if opts.SurvivesPivotRoot {
+		// to survive the pivot_root, we need to make the mount units depend on
+		// all of the various initrd special targets by adding runtime conf
+		// files there
+		// note we could do this statically in the initramfs main filesystem
+		// layout, but that means that changes to snap-bootstrap would block on
+		// waiting for those files to be added before things works here, this is
+		// a more flexible strategy that puts snap-bootstrap in control
+		for _, initrdUnit := range []string{
+			"initrd.target",
+			"initrd-fs.target",
+			"initrd-switch-root.target",
+		} {
```

Comment:
Note that currently ubuntu-core-initramfs as is the launchpad codebase does not perform isolation, and thus everything remains mounted.

But this PR, further down in testing, reverts that and goes back to isolating after daemon-reload, before re-exec. Previously that was considered fragile, as things that are setup from writable-paths were getting unmounted? I don't see how writable-paths are hooked up to initrd*.targets to ensure those too survive pivot-root with isolating to initrd.target.

But I still think we should make mount units to be wanted by the local-fs.target maybe? Cause then post-pivot root things will look normal too.

Reply:
As I recall, the issue was not around writable-paths getting unmounted, but rather that ubuntu-boot mount for run mode (and maybe recover mode too I don't remember) and ubuntu-data on /host for recover mode were not persisting past pivot_root. If there are issues with writable-path, I would imagine that our spread tests here in snapd would catch that with these changes and AFAICT none of them failed here so I think writable-path mounts are okay (see https://github.com/snapcore/snapd/blob/master/tests/core/writablepaths/task.yaml and https://github.com/snapcore/snapd/blob/master/tests/core/system-settings/task.yaml) (though perhaps there is some other hack I'm not aware of that makes writable-path mounts survive pivot_root). 

Noted about making them depend on local-fs.target too, it's cheap for us to do so and I don't see a reason not to, so let's do it.